From c6dfaaff81dba80b4cde5dd72e9131aec3126553 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Tue, 24 May 2011 14:36:31 -0600
Subject: [PATCH] Cache allAccessSections per project

The projectState object computed allAccessSections every
time they were needed to ensure cache coherency when rules
changed.  Cache a project's allAccessSections in the
ProjectState and keep track of the revisionSets of all the
relevant projects to be able to recalculate these when
access rules change.

Add some additonal tests to RefControlTest.

Change-Id: I872ceef359b715b2257348bdbedd12c0fb7a72f6
---
 .../gerrit/server/git/VersionedMetaData.java       |    7 +-
 .../google/gerrit/server/project/ProjectState.java |  221 +++++++++++++++++---
 .../gerrit/server/project/RefControlTest.java      |   78 ++++++--
 3 files changed, 261 insertions(+), 45 deletions(-)

diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/git/VersionedMetaData.java b/gerrit-server/src/main/java/com/google/gerrit/server/git/VersionedMetaData.java
index bb91044..2bd78d3 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/git/VersionedMetaData.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/git/VersionedMetaData.java
@@ -64,9 +64,12 @@ public abstract class VersionedMetaData {
   protected abstract void onSave(CommitBuilder commit) throws IOException,
       ConfigInvalidException;
 
-  /** @return revision of the metadata that was loaded. */
+  /** @return revision of the metadata that was loaded.  A null indicates
+   *          an undeterminable revision and should be considered different
+   *          from any revision, including other null revisions.
+   */
   public ObjectId getRevision() {
-    return revision.copy();
+    return revision == null ? null : revision.copy();
   }
 
   /** Initialize in-memory as though the repository branch doesn't exist. */
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/ProjectState.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/ProjectState.java
index 0d297cd..19cb135 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/ProjectState.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/ProjectState.java
@@ -28,6 +28,7 @@ import com.google.gerrit.server.git.ProjectConfig;
 import com.google.inject.Inject;
 import com.google.inject.assistedinject.Assisted;
 
+import org.eclipse.jgit.lib.ObjectId;
 import org.eclipse.jgit.lib.Ref;
 import org.eclipse.jgit.lib.Repository;
 
@@ -45,6 +46,123 @@ public class ProjectState {
     ProjectState create(ProjectConfig config);
   }
 
+  /** Used to record a snapshot of project's AccessSections along
+   *  with the revisionSet to which these sections apply
+   */
+  private static class AccessSections {
+    final Set<ObjectId> configRevisionSet;
+    final List<AccessSection> allAccessSections;
+
+    public AccessSections(final Set<ObjectId> configRevisionSet,
+        final List<AccessSection> allAccessSections) {
+      this.configRevisionSet = configRevisionSet;
+      this.allAccessSections = allAccessSections;
+    }
+  }
+
+  /** Read Only Version */
+  private static class ROAccessSections extends AccessSections {
+    public ROAccessSections(AccessSections as) {
+      super(as.configRevisionSet == null ? null :
+          Collections.unmodifiableSet(as.configRevisionSet),
+          Collections.unmodifiableList(as.allAccessSections));
+    }
+  }
+
+  /**
+   * Use this class to walk the ancestor tree, it will abort loops safely.
+   *
+   * This class provides the ability to store data of type T which
+   * can then be accessed with the get()/set() methods to pass
+   * to/from the child and parents, or to gather data between each
+   * parent invocation in the walkParent() method.
+   *
+   * This class is static to ensure that developers do not mistakingly
+   * access instance methods and variable in walkParent() since these
+   * would likely incorrectly always refer to the instance which
+   * initiated the walk and not the child ProjectState.
+   */
+  public abstract static class Walker<T> {
+    T data;
+    Set<Project.NameKey> seen = new HashSet<Project.NameKey>();
+
+    boolean abort = false;
+
+    final Project.NameKey wildProject;
+    final ProjectCache projectCache;
+
+    public Walker(Project.NameKey wild, ProjectCache cache) {
+      wildProject = wild;
+      projectCache = cache;
+    }
+
+    public T get() {
+      return data;
+    }
+
+    public void set(T d) {
+      data = d;
+    }
+
+    /**
+     * Used to indicate that the walk should no longer call any more
+     * parents at any level in the walk.  Use to quickly interrupt a
+     * walk.
+     */
+    public void abortWalk() {
+      abort = true;
+    }
+
+    /** Stack like walker (preserve/restore data before/after call) */
+    public T walkAncestor(ProjectState child, T d) {
+      T prev = get();
+      set(d);
+      T rtn = walkAncestor(child);
+      set(prev);
+      return rtn;
+    }
+
+    /**
+     * Call this method to walk each one of the parents and in turn
+     * call walkParent() with them as the parent argument.  Skips
+     * parents which have already been traversed by this walker
+     * at a previous level (to prevent looping).
+     */
+    public T walkAncestor(ProjectState child) {
+      if (! abort) {
+        Project.NameKey parent = child.getProject().getParent();
+        if (parent == null && ! child.isWildProject()) {
+          parent = wildProject;
+        }
+
+        if (seen.add(parent)) {
+          final ProjectState s = projectCache.get(parent);
+          if (s != null) {
+            walkParent(child, s);
+          }
+        }
+      }
+      return get();
+    }
+
+    /**
+     * In order to be able to walk all the way up the ancestory tree,
+     * this method must call a method in the parent ProjectState which
+     * will eventually call walkParents(...) somewhere in its code path.
+     */
+    public abstract void walkParent(ProjectState child, ProjectState parent);
+  }
+
+  /**
+   * Helper to setup the Walker with the required initiator's
+   *  instance variables
+   */
+  public abstract class MyWalker<T> extends Walker<T> {
+    public MyWalker() {
+      super(ProjectState.this.wildProject, ProjectState.this.projectCache);
+    }
+  }
+
   private final AnonymousUser anonymousUser;
   private final Project.NameKey wildProject;
   private final ProjectCache projectCache;
@@ -56,6 +174,7 @@ public class ProjectState {
 
   /** Last system time the configuration's revision was examined. */
   private transient long lastCheckTime;
+  private volatile AccessSections accessSections;
 
   @Inject
   protected ProjectState(final AnonymousUser anonymousUser,
@@ -124,48 +243,90 @@ public class ProjectState {
     return config;
   }
 
-  /** Get the rights that pertain only to this project. */
-  public Collection<AccessSection> getLocalAccessSections() {
-    return getConfig().getAccessSections();
-  }
+  /**
+   * @param walker Omit. Only used internally when recursing
+   * @return revision set of the metadata revisions of this project and
+   *         all its ancestors.  A null indicates an undeterminable
+   *         revision set and should be considered different from any
+   *         other revision set, including other null revision sets.
+   */
+  public Set<ObjectId> getConfigRevisionSet(Walker<Set<ObjectId>>... walker) {
+    if (walker.length == 0) {
+      return getConfigRevisionSet(this.new MyWalker<Set<ObjectId>>() {
+          {
+            set(new HashSet<ObjectId>());
+          }
+          @Override
+          public void walkParent(ProjectState child, ProjectState parent) {
+            parent.getConfigRevisionSet(this);
+          }
+        });
+    }
 
-  /** Get the rights this project inherits. */
-  public Collection<AccessSection> getInheritedAccessSections() {
-    if (isWildProject()) {
-      return Collections.emptyList();
+    ObjectId revision = config.getRevision();
+    if (revision == null) {
+      walker[0].set(null);
+      walker[0].abortWalk();
+      return null;
     }
+    walker[0].get().add(revision);
 
-    List<AccessSection> inherited = new ArrayList<AccessSection>();
-    Set<Project.NameKey> seen = new HashSet<Project.NameKey>();
-    Project.NameKey parent = getProject().getParent();
+    return walker[0].walkAncestor(this);
+  }
 
-    while (parent != null && seen.add(parent)) {
-      ProjectState s = projectCache.get(parent);
-      if (s != null) {
-        inherited.addAll(s.getLocalAccessSections());
-        parent = s.getProject().getParent();
-      } else {
-        break;
-      }
+  /** @param seen is the parents seen so far, used to detect loops */
+  protected AccessSections getAccessSections(Walker<AccessSections>... walker) {
+    Set<ObjectId> configRevisionSet = getConfigRevisionSet();
+    AccessSections as = accessSections;
+    if (as == null || configRevisionSet == null ||
+        !configRevisionSet.equals(as.configRevisionSet)) {
+      as = computeAccessSections(walker);
+      accessSections = as;
     }
+    return as;
+  }
 
-    // Wild project is the parent, or the root of the tree
-    if (parent == null) {
-      ProjectState s = projectCache.get(wildProject);
-      if (s != null) {
-        inherited.addAll(s.getLocalAccessSections());
-      }
+  /** @param walker Omit. Only used internally when recursing */
+  private ROAccessSections computeAccessSections(Walker<AccessSections>... walker) {
+    if (walker.length == 0) {
+      return computeAccessSections(this.new MyWalker<AccessSections>() {
+          @Override
+          public void walkParent(ProjectState child, ProjectState parent) {
+            AccessSections as = get();
+            AccessSections pas = parent.getAccessSections(this);
+            as.allAccessSections.addAll(pas.allAccessSections);
+            if (as.configRevisionSet != null && pas.configRevisionSet != null) {
+              as.configRevisionSet.addAll(pas.configRevisionSet);
+            } else {
+              set(new AccessSections(null, as.allAccessSections));
+            }
+          }
+        });
+    }
+
+    Set<ObjectId> revisions = null;
+    ObjectId revision = config.getRevision();
+    if (revision != null) {
+      revisions = new HashSet<ObjectId>();
+      revisions.add(revision);
     }
 
-    return inherited;
+    List<AccessSection> accessSections = new ArrayList<AccessSection>(
+        getConfig().getAccessSections());
+
+    AccessSections as = new AccessSections(revisions, accessSections);
+    walker[0].set(as);
+
+    if (! isWildProject()) {
+      walker[0].walkAncestor(this, as);
+    }
+    return new ROAccessSections(walker[0].get());
   }
 
+
   /** Get both local and inherited access sections. */
   public Collection<AccessSection> getAllAccessSections() {
-    List<AccessSection> all = new ArrayList<AccessSection>();
-    all.addAll(getLocalAccessSections());
-    all.addAll(getInheritedAccessSections());
-    return all;
+    return getAccessSections().allAccessSections;
   }
 
   /**
diff --git a/gerrit-server/src/test/java/com/google/gerrit/server/project/RefControlTest.java b/gerrit-server/src/test/java/com/google/gerrit/server/project/RefControlTest.java
index ab5d75f..3555d9e 100644
--- a/gerrit-server/src/test/java/com/google/gerrit/server/project/RefControlTest.java
+++ b/gerrit-server/src/test/java/com/google/gerrit/server/project/RefControlTest.java
@@ -76,13 +76,23 @@ public class RefControlTest extends TestCase {
 
     assertNotOwner("refs/*", uDev);
     assertNotOwner("refs/heads/master", uDev);
+    assertNotOwner("refs/heads/master/x", uDev);
+
+    // Odd stuff
+    assertNotOwner("refs/heads/*/x", uDev);
+    assertNotOwner("refs/*/x", uDev);
+    assertNotOwner("refs/heads/y/../x/*", uDev);
   }
 
   public void testBranchDelegation2() {
     grant(local, OWNER, admin, "refs/*");
     grant(local, OWNER, devs, "refs/heads/x/*");
     grant(local, OWNER, fixers, "refs/heads/x/y/*");
-    doNotInherit(local, OWNER, "refs/heads/x/y/*");
+    exclusive(local, OWNER, "refs/heads/x/y/*");
+
+    ProjectControl uAdmin = user(devs, admin);
+    assertTrue("owns ref", uAdmin.isOwner());
+    assertOwner("refs/heads/x/y/*", uAdmin);
 
     ProjectControl uDev = user(devs);
     assertFalse("not owner", uDev.isOwner());
@@ -106,12 +116,54 @@ public class RefControlTest extends TestCase {
     assertNotOwner("refs/heads/master", uFix);
   }
 
+  public void testInherit() {
+    grant(wild, READ, registered, "refs/*");
+    grant(parent, PUSH, registered, "refs/*");
+
+    ProjectControl u = user();
+    assertTrue("can read", u.isVisible());
+    assertTrue("can push", u.canPushToAtLeastOneRef());
+  }
+
+  public void testInheritRead_OverrideWithDeny() {
+    grant(parent, READ, registered, "refs/*");
+    grant(local, READ, registered, "refs/*").setDeny(true);
+
+    ProjectControl u = user();
+    assertFalse("can't read", u.isVisible());
+  }
+
+  public void testInheritRead_OverrideWithDeny2() {
+    grant(wild, READ, registered, "refs/*");
+    grant(local, READ, registered, "refs/*").setDeny(true);
+
+    ProjectControl u = user();
+    assertFalse("can't read", u.isVisible());
+  }
+
+  public void testInheritRead_OverrideWithDeny3() {
+    grant(wild, READ, registered, "refs/*");
+    grant(parent, READ, registered, "refs/*").setDeny(true);
+
+    ProjectControl u = user();
+    assertFalse("can't read", u.isVisible());
+  }
+
+  public void testInheritRead_RevertOverride() {
+    grant(wild, READ, registered, "refs/*");
+    grant(parent, READ, registered, "refs/*").setDeny(true);
+    grant(local, READ, registered, "refs/*");
+
+    ProjectControl u = user();
+    assertTrue("can read", u.isVisible());
+  }
+
   public void testInheritRead_SingleBranchDeniesUpload() {
     grant(parent, READ, registered, "refs/*");
     grant(parent, PUSH, registered, "refs/for/refs/*");
     grant(local, READ, registered, "refs/heads/foobar");
-    doNotInherit(local, READ, "refs/heads/foobar");
-    doNotInherit(local, PUSH, "refs/for/refs/heads/foobar");
+    exclusive(local, READ, "refs/heads/foobar");
+    exclusive(local, PUSH, "refs/for/refs/heads/foobar");
 
     ProjectControl u = user();
     assertTrue("can upload", u.canPushToAtLeastOneRef());
@@ -138,14 +190,6 @@ public class RefControlTest extends TestCase {
         u.controlForRef("refs/heads/foobar").canUpload());
   }
 
-  public void testInheritRead_OverrideWithDeny() {
-    grant(parent, READ, registered, "refs/*");
-    grant(local, READ, registered, "refs/*").setDeny(true);
-
-    ProjectControl u = user();
-    assertFalse("can't read", u.isVisible());
-  }
-
   public void testInheritRead_AppendWithDenyOfRef() {
     grant(parent, READ, registered, "refs/*");
     grant(local, READ, registered, "refs/heads/*").setDeny(true);
@@ -194,6 +238,9 @@ public class RefControlTest extends TestCase {
 
   // -----------------------------------------------------------------------
 
+  private Project.NameKey wildProject = new Project.NameKey("-- All Projects --");
+
+  private ProjectConfig wild;
   private ProjectConfig local;
   private ProjectConfig parent;
   private final AccountGroup.UUID admin = new AccountGroup.UUID("test.admin");
@@ -238,6 +285,9 @@ public class RefControlTest extends TestCase {
   public void setUp() throws Exception {
     super.setUp();
 
+    wild = new ProjectConfig(wildProject);
+    wild.createInMemory();
+
     parent = new ProjectConfig(new Project.NameKey("parent"));
     parent.createInMemory();
 
@@ -263,7 +313,7 @@ public class RefControlTest extends TestCase {
     return rule;
   }
 
-  private void doNotInherit(ProjectConfig project, String permissionName,
+  private void exclusive(ProjectConfig project, String permissionName,
       String ref) {
     project.getAccessSection(ref, true) //
         .getPermission(permissionName, true) //
@@ -318,12 +368,14 @@ public class RefControlTest extends TestCase {
     };
 
     GitRepositoryManager mgr = null;
-    Project.NameKey wildProject = new Project.NameKey("-- All Projects --");
     ProjectControl.AssistedFactory projectControlFactory = null;
     all.put(local.getProject().getNameKey(), new ProjectState(anonymousUser,
         projectCache, wildProject, projectControlFactory, mgr, local));
     all.put(parent.getProject().getNameKey(), new ProjectState(anonymousUser,
         projectCache, wildProject, projectControlFactory, mgr, parent));
+    all.put(wild.getProject().getNameKey(), new ProjectState(anonymousUser,
+        projectCache, wildProject, projectControlFactory, mgr, wild));
+
     return all.get(local.getProject().getNameKey());
   }
 
-- 
1.7.4.1

